lib/events.ts

```diff
@@ -0,0 +1,122 @@
+// deno-lint-ignore-file no-explicit-any ban-types
+import { createChannel } from "./channel.ts";
+import { action, resource, suspend } from "./instructions.ts";
+import { useScope } from "./run/scope.ts";
+import type { Operation, Stream } from "./types.ts";
+
+type FN = (...any: any[]) => any;
+
+type EventTypeFromEventTarget<T, K extends string> = `on${K}` extends keyof T
+  ? Parameters<Extract<T[`on${K}`], FN>>[0]
+  : Event;
+
+export type EventList<T> = T extends {
+  addEventListener(type: infer P, ...args: any): void;
+  // we basically ignore this but we need it so we always get the first override of addEventListener
+  addEventListener(type: infer P2, ...args: any): void;
+} ? P & string
+  : never;
+
+/**
+ * Takes an event source and event name and returns an
+ * {@link effection:Operation} that produces the value emitted by that
+ * event. Both {@link EventTarget} and {@link EventEmitter} are
+ * supported.
+ *
+ * ### Example
+ *
+ * ``` javascript
+ * let start = yield once(document, 'dragstart');
+ * console.log(`drag started at (${start.pageX}, ${start.pageY})`);
+ *
+ * let end = yield once(document, 'dragend');
+ * console.log(`drag ended at (${end.pageX}, ${end.pageY})`);
+ * ```
+ *
+ * ### Example
+ *
+ * ``` javascript
+ * let src = yield once(stream, 'pipe');
+ * ```
+ *
+ * @param source an object which emits events
+ * @param eventName the name of the event to subscribe to
+ * @typeParam T the type of the argument to the emitted event
+ */
+export function once<
+  T extends EventTarget,
+  K extends EventList<T> | (string & {}),
+>(target: T, name: K): Operation<EventTypeFromEventTarget<T, K>> {
+  return action(function* (resolve) {
+    target.addEventListener(
+      name,
+      resolve as EventListenerOrEventListenerObject,
+    );
+    try {
+      yield* suspend();
+    } finally {
+      target.removeEventListener(
+        name,
+        resolve as EventListenerOrEventListenerObject,
+      );
+    }
+  });
+}
```

Comment:
Can we implement this using [`first()`] on the stream? https://github.com/thefrontside/effection/blob/v3/lib/first.ts#L3